src/rust/catloop/futures/connect.rs

```diff
@@ -46,288 +47,192 @@ use ::std::{
-            // We may get some error.
-            demi_opcode_t::DEMI_OPC_FAILED => {
-                let cause: String = format!(
-                    "failed to establish connection (qd={:?}, qt={:?}, errno={:?})",
-                    qr.qr_qd, qt_rx, qr.qr_ret
-                );
-                error!("connect_ack_received(): {:?}", &cause);
-                return Poll::Ready(Err(Fail::new(qr.qr_ret as i32, &cause)));
-            },
-            // We do not expect anything else.
-            _ => {
-                // The following statement is unreachable because we have issued a pop operation.
-                // If we successfully complete a different operation, something really bad happen in the scheduler.
-                unreachable!("unexpected operation on control duplex pipe")
-            },
+async fn send_ack(catmem: Rc<RefCell<CatmemLibOS>>, new_pipe: Rc<DuplexPipe>, yielder: &Yielder) -> Result<(), Fail> {
+    // Create message with magic connect.
+    let sga: demi_sgarray_t = CatloopLibOS::cook_magic_connect(&catmem)?;
+    // Send to server through new pipe.
+    let qt: QToken = new_pipe.push(&sga)?;
+    trace!("Send ack qtoken={:?}", qt);
+    // Get scheduler handle from the task id.
+    let mut catmem_: RefMut<CatmemLibOS> = catmem.borrow_mut();
+    let handle: TaskHandle = catmem_.schedule(qt)?;
+    // Drop the mutable reference because we might yield.
+    drop(catmem_);
+
+    // Yield until push completes.
+    while !handle.has_completed() {
+        if let Err(e) = yielder.yield_once().await {
+            return Err(e);
         }
-
-        // Extract scatter-gather array from operation result.
-        let sga: demi_sgarray_t = unsafe { qr.qr_value.sga };
-
-        // Extract port number.
-        let port: u16 = {
-            let port: Result<u16, Fail> = extract_port_number(&sga);
-            self_.catmem.borrow_mut().free_sgarray(sga)?;
-            self_.control_duplex_pipe.shutdown()?;
-            port?
-        };
-
-        // Open underlying pipes.
-        let remote: SocketAddrV4 = SocketAddrV4::new(self_.ipv4, port);
-        let duplex_pipe: Rc<DuplexPipe> =
-            Rc::new(DuplexPipe::open_duplex_pipe(self_.catmem.clone(), &self_.ipv4, port)?);
-
-        let sga: demi_sgarray_t = CatloopLibOS::cook_magic_connect(&self_.catmem)?;
-        let qt_tx: QToken = duplex_pipe.push(&sga)?;
-        self_.catmem.borrow_mut().free_sgarray(sga)?;
-
-        // Transition to the next state in the connection establishment protocol.
-        self_.state = ClientState::Connected {
-            qt_tx,
-            remote,
-            duplex_pipe,
-        };
-    } else {
-        if attempt > MAX_ACK_RECEIVED_ATTEMPTS {
-            // Connection timeout, retry.
-            debug!(
-                "connect_ack_received(): connection timeout, retrying (qt_rx={:?})",
-                qt_rx
+    }
+    // Re-acquire reference to catmem libos.
+    catmem_ = catmem.borrow_mut();
```

Comment:
What if is not possible to reborrow? This may happend if some concurrent, buggy async function yields without calling drop. Re-acquire a borrow is NOT like re-acquiring a lock. The former will cause a runtime panic, whereas the latter will spin. We should consider using `try_borrow_mut()`.